zkvm/src/types.rs

```diff
@@ -136,6 +139,43 @@ impl ScalarWitness {
     }
 }
 
+impl Neg for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn neg(self) -> ScalarWitness {
+        match self {
+            ScalarWitness::Integer(a) => ScalarWitness::Integer(-a),
+            ScalarWitness::Scalar(a) => ScalarWitness::Scalar(-a),
+        }
+    }
+}
+impl Add for ScalarWitness {
+    type Output = ScalarWitness;
+
+    fn add(self, rhs: ScalarWitness) -> ScalarWitness {
+        match (self, rhs) {
+            (ScalarWitness::Integer(a), ScalarWitness::Integer(b)) => {
+                let res = a + b;
+
+                if res > spacesuit::SignedInteger::from(std::u64::MAX) {
```

Comment:
oh, that's clever use of capped i128. But the result can also be lower than negated u64::MAX, so we need to check that too:

```suggestion
                let max = spacesuit::SignedInteger::from(std::u64::MAX);
                let min = -max;
                if res > max || res < min {
```